### PR TITLE
feat: [CI-5820]: Added pipeline save support for simplified yaml

### PR DESCRIPTION
**Pull request by raghavharness**
## Describe your changes
Added pipeline save support for simplified yaml
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38844)

### CODE REVIEW

File: pipeline-service/service/src/main/java/io/harness/pms/pipeline/mappers/PMSPipelineFilterHelper.java [hunk 1]
```diff
@@ -29,6 +29,7 @@ public Update getUpdateOperations(PipelineEntity pipelineEntity, long timestamp)
     update.set(PipelineEntityKeys.stageCount, pipelineEntity.getStageCount());
     update.set(PipelineEntityKeys.stageNames, pipelineEntity.getStageNames());
     update.set(PipelineEntityKeys.allowStageExecutions, pipelineEntity.getAllowStageExecutions());
+    update.set(PipelineEntityKeys.harnessVersion, pipelineEntity.getHarnessVersion());
```
Comment (raghavharness) [inline pipeline-service/service/src/main/java/io/harness/pms/pipeline/mappers/PMSPipelineFilterHelper.java:32]: @hextechpal I have added this change to populate V0 to existing pipeline updates. Let me know if this is fine.

**Comment by raghavharness** [comment]
We can ignore the PMD check, it has been deprecated.
https://pmd.github.io/latest/pmd_rules_java_performance.html#simplifystartswith

**Review by hextechpal — CHANGES_REQUESTED** [review]
> Reviewed 11 of 16 files at r2, 2 of 5 files at r3, 2 of 3 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38844)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java` line 468 at r5](https://reviewable.io/reviews/harness/harness-core/38844#-NEdhiKR4e_ZCqsPAVkX:-NEdhiKR4e_ZCqsPAVkY:bv8fl2u) ([raw file](https://github.com/harness/harness-core/blob/29468e70dada471aec1528e72d61ff0942a9c415/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L468)):*
> ```Java
>       "This feature flag helps in synchronizing the git clone of repo and copying the files then to destination directory",
>       HarnessTeam.CDP),
>   CI_ENABLE_YAML_SIMPLIFICATION("Feature flag for yaml simplification", HarnessTeam.CI);
> ```

YAML_VERSIONING probably is a better FF name
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/mappers/PMSPipelineFilterHelper.java` line 32 at r2](https://reviewable.io/reviews/harness/harness-core/38844#-NEaJiCSq-GNVqeDdNzH-r2-32:-NEdfSH02HLv50Y5VyhJ:b-pysbbv) ([raw file](https://github.com/harness/harness-core/blob/4a7a43f485be0d2d1a25878a71574fef40cb1343/pipeline-service/service/src/main/java/io/harness/pms/pipeline/mappers/PMSPipelineFilterHelper.java#L32)):*
<details><summary><i>Previously, raghavharness (Raghav) wrote…</i></summary><blockquote>

@hextechpal I have added this change to populate V0 to existing pipeline updates. Let me know if this is fine. 
</blockquote></details>

I think it is fine
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceHelper.java` line 149 at r5](https://reviewable.io/reviews/harness/harness-core/38844#-NEdfwVEBAOH709Pxxk0:-NEdfwVEBAOH709Pxxk1:b12jx9r) ([raw file](https://github.com/harness/harness-core/blob/29468e70dada471aec1528e72d61ff0942a9c415/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceHelper.java#L149)):*
> ```Java
>   public PipelineEntity updatePipelineInfo(PipelineEntity pipelineEntity, YamlVersion pipelineVersion)
>       throws IOException {
>     if (pipelineVersion != null && pipelineVersion != YamlVersion.V0) {
> ```

create a switch case here and pipeline version should never be null. so check is redundant
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceHelper.java` line 443 at r5](https://reviewable.io/reviews/harness/harness-core/38844#-NEdg7557ZB3fbZcuRg_:-NEdg7557ZB3fbZcuRga:b-7ropln) ([raw file](https://github.com/harness/harness-core/blob/29468e70dada471aec1528e72d61ff0942a9c415/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceHelper.java#L443)):*
> ```Java
>   public static void checkAndThrowMismatchInImportedPipelineMetadata(String orgIdentifier, String projectIdentifier,
>       String pipelineIdentifier, PipelineImportRequestDTO pipelineImportRequest, String importedPipeline,
>       YamlVersion pipelineVersion) {
> ```

ditto
___
*[`pipeline-service/service/src/test/java/io/harness/pms/pipeline/service/PipelineServiceFormCriteriaTest.java` line 106 at r5](https://reviewable.io/reviews/harness/harness-core/38844#-NEdhRh4Ey-88IbYLTtE:-NEdhRh4Ey-88IbYLTtF:b-ic9uji) ([raw file](https://github.com/harness/harness-core/blob/29468e70dada471aec1528e72d61ff0942a9c415/pipeline-service/service/src/test/java/io/harness/pms/pipeline/service/PipelineServiceFormCriteriaTest.java#L106)):*
> ```Java
>     on(pmsPipelineService).set("pmsPipelineRepository", pmsPipelineRepository);
>     doReturn(outboxEvent).when(outboxService).save(any());
>     doReturn(updatedPipelineEntity).when(pmsPipelineServiceHelperMocked).updatePipelineInfo(pipelineEntity, null);
> ```

pass version null should never be the case

**Review by hextechpal — DISMISSED** [review]
> Reviewed 5 of 6 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38844)* status: :shipit: complete! ✓

**Review by hextechpal — DISMISSED** [review]
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38844)* status: :shipit: complete! ✓

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 16 files at r1, 7 of 16 files at r2, 2 of 5 files at r3, 1 of 3 files at r4, 5 of 6 files at r6, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38844)* status: :shipit: complete! ✓

___
*[`970-ng-commons/src/main/java/io/harness/pms/utils/PmsUtils.java` line 22 at r8](https://reviewable.io/reviews/harness/harness-core/38844#-NEeojhF0K8FgcqhM1nU:-NEeojhF0K8FgcqhM1nV:b-smwula) ([raw file](https://github.com/harness/harness-core/blob/75965f9f4760aa2ccf0885388a06e58f53628599/970-ng-commons/src/main/java/io/harness/pms/utils/PmsUtils.java#L22)):*
> ```Java
> @OwnedBy(PIPELINE)
> @UtilityClass
> public class PmsUtils {
> ```

Such pms utils can we move from ng-commons to 884-pms-commons ?
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java` line 627 at r8](https://reviewable.io/reviews/harness/harness-core/38844#-NEepN3o8dtmBUZ4aSLs:-NEepN3o8dtmBUZ4aSLt:b4u8w5v) ([raw file](https://github.com/harness/harness-core/blob/75965f9f4760aa2ccf0885388a06e58f53628599/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java#L627)):*
> ```Java
> 
>     if (isYamlSimplificationEnabled && version != null) {
>       assert version.getNode() != null;
> ```

Instead of such assert cant we add proper exception throwing by checking if its not null



_Code quote:_
```Java
assert version.getNode() != null;
```
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/yaml/PipelineYaml.java` line 28 at r8](https://reviewable.io/reviews/harness/harness-core/38844#-NEepocu8xLUiQICSbxr:-NEepocu8xLUiQICSbxs:b-3s8fcr) ([raw file](https://github.com/harness/harness-core/blob/75965f9f4760aa2ccf0885388a06e58f53628599/pipeline-service/service/src/main/java/io/harness/pms/pipeline/yaml/PipelineYaml.java#L28)):*
> ```Java
>   @JsonProperty("version") String version;
>   @JsonProperty("name") @EntityName String name;
> }
> ```

Will it not include stages etc?

**Comment by raghavharness** [comment]
@archit-harness 

> Such pms utils can we move from ng-commons to 884-pms-commons ?
Discussed with Prashant, changed it's name but keeping it to ng-commons since it could be used by other modules.

> Will it not include stages etc?
Right now we just need these 2, we will add if required in subsequent PR's

**Review by archit-harness — COMMENTED** [review]
> Reviewed 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38844)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java` line 627 at r8](https://reviewable.io/reviews/harness/harness-core/38844#-NEepN3o8dtmBUZ4aSLs:-NEfKByu6j2FxqWVzU8O:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/75965f9f4760aa2ccf0885388a06e58f53628599/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineServiceImpl.java#L627)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Instead of such assert cant we add proper exception throwing by checking if its not null
</blockquote></details>

done
___
*[`970-ng-commons/src/main/java/io/harness/pms/utils/PmsUtils.java` line 22 at r8](https://reviewable.io/reviews/harness/harness-core/38844#-NEeojhF0K8FgcqhM1nU:-NEfK8uL2TELNEp5ZJ0_:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/75965f9f4760aa2ccf0885388a06e58f53628599/970-ng-commons/src/main/java/io/harness/pms/utils/PmsUtils.java#L22)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Such pms utils can we move from ng-commons to 884-pms-commons ?
</blockquote></details>

done

**Review by archit-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38844)* status: :shipit: complete! ✓

___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/yaml/PipelineYaml.java` line 28 at r8](https://reviewable.io/reviews/harness/harness-core/38844#-NEepocu8xLUiQICSbxr:-NEfKn8lEEemQVdCtdau:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/75965f9f4760aa2ccf0885388a06e58f53628599/pipeline-service/service/src/main/java/io/harness/pms/pipeline/yaml/PipelineYaml.java#L28)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Will it not include stages etc?
</blockquote></details>

done

**Pull request merged**